packages/editor/src/lib/editor/tools/StateNode.ts

```diff
@@ -129,39 +130,138 @@ export abstract class StateNode implements Partial<TLEventHandlers> {
 	}
 	private _isActive: Atom<boolean>
 
+	getDescendant<T extends StateNode>(path: string): T | undefined {
+		const ids = path.split('.').reverse()
+		let state = this as any
+		while (ids.length > 0) {
+			const id = ids.pop()
+			if (!id) return state as T
+			const childState = state.children?.[id]
+			if (!childState) return undefined
+			state = childState
+		}
+		return state as T
+	}
+
+	private exitBranch(id: string, info: any = {}) {
+		let nextExitingParent = this as StateNode | undefined
+		while (nextExitingParent) {
```

Comment:
a comment would be good for this bit of logic, hurts my brain a bit

Reply:
i think it's because i usually think of exiting the branch upwards,  it exits the parent, but then the logic references the `nextExitingChild` so it makes it look like it's going downwards 🙃

Reply:
I believe that's right, we're disabling the branch here downwards (parent > child) when we exit an active branch. I don't know whether there's a strong convention here, but we could try either way and see if any tests fail?